sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala

```diff
@@ -40,13 +40,14 @@ import org.apache.spark.sql.types._
  * requested.  The attributes produced by this function will be automatically copied anytime rules
  * result in changes to the Generator or its children.
  */
-abstract class Generator extends Expression {
-  self: Product =>
+trait Generator extends Expression { self: Product =>
```

Comment:
Nothing. Unless you have circular dependencies between traits, which I doubt. I avoid self-types as much as I can. Inheritance is much easier to understand and covers 99% of cases.